File: mobile/packages/divine_video_player/android/src/main/kotlin/com/divinevideo/divine_video_player/DivineVideoPlayerLog.kt
// ABOUTME: Forwards curated native video-player diagnostics to Dart's
// ABOUTME: UnifiedLogger so playback issues appear in user bug reports

package com.divinevideo.divine_video_player

import android.util.Log

/**
 * Bridges relevant native diagnostics to the Dart side so they are captured by
 * the app's `UnifiedLogger` and included in bug-report log dumps.
 *
 * Native playback code calls these methods for the handful of events worth
 * surfacing to support: player lifecycle, clip-composition skips, asset /
 * playback errors, and audio-track setup. Per-frame / verbose logging must stay
 * on [Log] so it does not flood the captured buffer.
 *
 * `DivineVideoPlayerPlugin` installs [sink] (only for the main FlutterEngine)
 * to forward each entry over the global method channel. Before that is wired
 * (or in unit-test contexts) entries fall back to logcat only.
 */
object DivineVideoPlayerLog {
    /**
     * Forwards `(level, message, name)` to Dart. `level` is one of `debug`,
     * `info`, `warning`, `error`. Set by the plugin; `null` until then.
     */
    @Volatile
    var sink: ((String, String, String) -> Unit)? = null

    fun debug(message: String, name: String = "DivineVideoPlayer") = emit("debug", message, name)

    fun info(message: String, name: String = "DivineVideoPlayer") = emit("info", message, name)

    fun warning(message: String, name: String = "DivineVideoPlayer") = emit("warning", message, name)

    fun error(message: String, name: String = "DivineVideoPlayer") = emit("error", message, name)

    private fun emit(level: String, message: String, name: String) {
        // Keep the logcat fallback so on-device debugging is unchanged.
        when (level) {
            "error" -> Log.e(name, message)
            "warning" -> Log.w(name, message)
            else -> Log.d(name, message)
        }
        sink?.invoke(level, message, name)
    }
}

File: mobile/packages/divine_video_player/ios/Classes/DivineVideoPlayerLog.swift
// ABOUTME: Forwards curated native video-player diagnostics to Dart's
// ABOUTME: UnifiedLogger so playback issues appear in user bug reports

import Foundation

/// Bridges relevant native diagnostics to the Dart side so they are captured
/// by the app's `UnifiedLogger` and included in bug-report log dumps.
///
/// Native playback code calls these methods for the handful of events worth
/// surfacing to support: player lifecycle, clip-composition skips, asset /
/// playback errors, and audio-track setup. Per-frame / verbose logging must
/// stay on `print` so it does not flood the captured buffer.
///
/// `DivineVideoPlayerPlugin` installs `sink` to forward each entry over the
/// global method channel. Before that is wired (or in unit/host contexts)
/// entries fall back to the console only.
final class DivineVideoPlayerLog {
    static let shared = DivineVideoPlayerLog()

    private init() {}

    /// Forwards `(level, message, name)` to Dart. `level` is one of
    /// `debug`, `info`, `warning`, `error`. Set by the plugin; `nil` until then.
    var sink: ((String, String, String) -> Void)?

    func debug(_ message: String, name: String = "DivineVideoPlayer") {
        emit("debug", message, name)
    }

    func info(_ message: String, name: String = "DivineVideoPlayer") {
        emit("info", message, name)
    }

    func warning(_ message: String, name: String = "DivineVideoPlayer") {
        emit("warning", message, name)
    }

    func error(_ message: String, name: String = "DivineVideoPlayer") {
        emit("error", message, name)
    }

    private func emit(_ level: String, _ message: String, _ name: String) {
        // Keep the console fallback so on-device debugging is unchanged.
        print("[\(name)] \(message)")
        sink?(level, message, name)
    }
}
